### PR TITLE
feat(ruleset-migrator): implement ruleset migrator 

### PR DESCRIPTION
**Pull request by P0lip**
Closes https://github.com/stoplightio/platform-internal/issues/6763

**Checklist**

- [x] Tests added / updated
- [x] Docs added / updated

**Does this PR introduce a breaking change?**

- [ ] Yes
HonestWork

- [x] No

**Additional context**

Make sure to read https://github.com/stoplightio/spectral/pull/1698/commits/f90d2c8cdd404865db67d56eea093b735af65637

Tests account for ~50% of the whole diff

```
 .../services/__tests__/__fixtures__/ruleset.json   |  17 +++
 packages/cli/src/services/__tests__/linter.test.ts |  14 ++
 .../__fixtures__/apisyouwonthate/output.cjs        | 149 +++++++++++++++++++++
 .../__fixtures__/apisyouwonthate/output.mjs        | 143 ++++++++++++++++++++
 .../__fixtures__/apisyouwonthate/ruleset.yaml      | 133 ++++++++++++++++++
 .../__fixtures__/extends-variant-1/output.cjs      |   4 +
 .../__fixtures__/extends-variant-1/output.mjs      |   4 +
 .../__fixtures__/extends-variant-1/ruleset.yaml    |   1 +
 .../__fixtures__/extends-variant-2/output.cjs      |   4 +
 .../__fixtures__/extends-variant-2/output.mjs      |   4 +
 .../__fixtures__/extends-variant-2/ruleset.yaml    |   1 +
 .../__fixtures__/extends-variant-3/output.cjs      |   4 +
 .../__fixtures__/extends-variant-3/output.mjs      |   4 +
 .../__fixtures__/extends-variant-3/ruleset.yaml    |   1 +
 .../__fixtures__/extends-variant-4/output.cjs      |   4 +
 .../__fixtures__/extends-variant-4/output.mjs      |   4 +
 .../__fixtures__/extends-variant-4/ruleset.yaml    |   1 +
 .../__fixtures__/extends-variant-5/output.cjs      |   4 +
 .../__fixtures__/extends-variant-5/output.mjs      |   4 +
 .../__fixtures__/extends-variant-5/ruleset.yaml    |   1 +
 .../__fixtures__/formats-variant-1/output.cjs      |  10 ++
 .../__fixtures__/formats-variant-1/output.mjs      |   5 +
 .../__fixtures__/formats-variant-1/ruleset.yaml    |   2 +
 .../__fixtures__/formats-variant-2/output.cjs      |  21 +++
 .../__fixtures__/formats-variant-2/output.mjs      |  14 ++
 .../__fixtures__/formats-variant-2/ruleset.yaml    |  16 +++
 .../__fixtures__/functions-variant-1/output.cjs    |  28 ++++
 .../__fixtures__/functions-variant-1/output.mjs    |  25 ++++
 .../__fixtures__/functions-variant-1/ruleset.yaml  |  22 +++
 .../__fixtures__/functions-variant-2/output.cjs    |  23 ++++
 .../__fixtures__/functions-variant-2/output.mjs    |  20 +++
 .../__fixtures__/functions-variant-2/ruleset.yaml  |  17 +++
 .../__fixtures__/rules-variant-1/output.cjs        |  19 +++
 .../__fixtures__/rules-variant-1/output.mjs        |  19 +++
 .../__fixtures__/rules-variant-1/ruleset.yaml      |  16 +++
 .../ruleset-migrator/src/__tests__/ruleset.test.ts |  63 +++++++++
 36 files changed, 821 insertions(+)
```

One can apply the following filter to filter out the majority of test artefacts.

image

### CODE REVIEW

File: packages/cli/src/services/linter/utils/getRuleset.ts [hunk 1]
```diff
@@ -2,8 +2,12 @@ import { Optional } from '@stoplight/types';
 import { Ruleset, RulesetDefinition } from '@stoplight/spectral-core';
 import * as fs from 'fs';
 import * as path from '@stoplight/path';
-import { isAbsolute } from '@stoplight/path';
 import * as process from 'process';
+import { extname } from '@stoplight/path';
+import { migrateRuleset } from '@stoplight/spectral-ruleset-migrator';
+
+// eslint-disable-next-line @typescript-eslint/require-await
+const AsyncFunction = (async (): Promise<void> => void 0).constructor as FunctionConstructor;
```
Comment (wmhilton) [inline packages/cli/src/services/linter/utils/getRuleset.ts:10]: Do I even want to know?

Comment (P0lip) [reply packages/cli/src/services/linter/utils/getRuleset.ts:10]: So, long story short - in rare circumstances a "top-level" await might be inserted.
This won't be a problem soon-ish once we hop on ESM, but for the time being it's a plain CommonJS.
The above is needed for us to be able to do

```
AsyncFunction('module.exports = await /* the rest of the code goes here */)'
```

Comment (wmhilton) [reply packages/cli/src/services/linter/utils/getRuleset.ts:10]: Narrator: it turns out he really did not want to know :trollface:

**Review by billiegoose — DISMISSED** [review]
Very nice!

**Pull request merged**